[CP] Refactor CP sharding rules into separate module and only register when CP is enabled

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* __->__ #167381

Previously, CP-specific sharding strategies (which shard on the sequence dimension) were directly
included in the base sharding strategies for scaled_dot_product_attention operators in
`_matrix_ops.py`. This meant these strategies were always available, even when CP was not enabled,
which could lead to incorrect sharding behavior as these sharding rules are not mathmetically correct without CP.

1. **Created new module**:
`torch/distributed/tensor/experimental/_context_parallel/_sharding_rules.py`
   - Implements `op_strategy_context()` - a context manager for temporarily
registering/unregistering strategies
   - Defines CP-enhanced strategy functions for all 6 scaled_dot_product_attention ops (forward and
 backward for flash, efficient, and cudnn variants)
   - Provides `register_cp_sharding_rules()` and `unregister_cp_sharding_rules()` APIs

2. **Updated `_matrix_ops.py`**
   - Removed all CP-specific sharding rules (sequence dimension sharding strategies)
   - Base strategies now only contain replicate, tensor parallelism, and batch sharding
strategies

3. **Updated `_attention.py`**
   - `_enable_cp_dtensor_dispatcher()` now calls `register_cp_sharding_rules()` to dynamically add
CP strategies
   - ~`_disable_cp_dtensor_dispatcher()` now calls `unregister_cp_sharding_rules()` to restore
original strategies~ This will invalidate all the sharding prop caches. Disable it for now.

cc @H-Huang @awgu @wanchaol @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

tianyu-l commented on torch/distributed/tensor/experimental/_context_parallel/_sharding_rules.py:
```diff
@@ -0,0 +1,387 @@
+
+
+@contextmanager
+def op_strategy_context(op_overload, strategy_func, schema_info=None):
```
I feel a context manager is not absolute necessary. Is it just for convenience?

fegin committed:
Update

[ghstack-poisoned]

fegin commented on torch/distributed/tensor/experimental/_context_parallel/_sharding_rules.py:
```diff
@@ -0,0 +1,387 @@
+    )
+
+    # Append CP strategies to base strategies
+    base_strategy.strategies.extend(cp_strategies.strategies)
```
ye, I think you are right. Here is the proposal to do the correct one.

1. We extract the core logic of  scaled_dot_product_flash_attention_strategy to a utility function. But the utility function doesn't call expand_to_full_mesh_op_strategy.
2. The original scaled_dot_product_flash_attention_strategy call the utility function and then call expand_to_full_mesh_op_strategy
3. scaled_dot_product_flash_attention_cp_strategy call the utility function and then append the Shard(2) 1D placement and then call expand_to_full_mesh_op_strategy.

So again, all attentions will be Shard(2) aware if CP is enabled.

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update on "[CP] Refactor CP sharding rules into separate module and only register when CP is enabled"


Previously, CP-specific sharding strategies (which shard on the sequence dimension) were directly
included in the base sharding strategies for scaled_dot_product_attention operators in
`_matrix_ops.py`. This meant these strategies were always available, even when CP was not enabled,
which could lead to incorrect sharding behavior as these sharding rules are not mathmetically correct without CP.

1. **Created new module**:
`torch/distributed/tensor/experimental/_context_parallel/_sharding_rules.py`
   - Implements `op_strategy_context()` - a context manager for temporarily
registering/unregistering strategies
   - Defines CP-enhanced strategy functions for all 6 scaled_dot_product_attention ops (forward and
 backward for flash, efficient, and cudnn variants)
   - Provides `register_cp_sharding_rules()` and `unregister_cp_sharding_rules()` APIs

2. **Updated `_matrix_ops.py`**
   - Removed all CP-specific sharding rules (sequence dimension sharding strategies)
   - Base strategies now only contain replicate, tensor parallelism, and batch sharding
strategies

3. **Updated `_attention.py`**
   - `_enable_cp_dtensor_dispatcher()` now calls `register_cp_sharding_rules()` to dynamically add
CP strategies
   - ~`_disable_cp_dtensor_dispatcher()` now calls `unregister_cp_sharding_rules()` to restore
original strategies~ This will invalidate all the sharding prop caches. Disable it for now.

cc H-Huang awgu wanchaol fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

fegin committed:
Update on "[CP] Refactor CP sharding rules into separate module and only register when CP is enabled"


Previously, CP-specific sharding strategies (which shard on the sequence dimension) were directly
included in the base sharding strategies for scaled_dot_product_attention operators in
`_matrix_ops.py`. This meant these strategies were always available, even when CP was not enabled,
which could lead to incorrect sharding behavior as these sharding rules are not mathmetically correct without CP.

1. **Created new module**:
`torch/distributed/tensor/experimental/_context_parallel/_sharding_rules.py`
   - Implements `op_strategy_context()` - a context manager for temporarily
registering/unregistering strategies
   - Defines CP-enhanced strategy functions for all 6 scaled_dot_product_attention ops (forward and
 backward for flash, efficient, and cudnn variants)
   - Provides `register_cp_sharding_rules()` and `unregister_cp_sharding_rules()` APIs

2. **Updated `_matrix_ops.py`**
   - Removed all CP-specific sharding rules (sequence dimension sharding strategies)
   - Base strategies now only contain replicate, tensor parallelism, and batch sharding
strategies

3. **Updated `_attention.py`**
   - `_enable_cp_dtensor_dispatcher()` now calls `register_cp_sharding_rules()` to dynamically add
CP strategies
   - ~`_disable_cp_dtensor_dispatcher()` now calls `unregister_cp_sharding_rules()` to restore
original strategies~ This will invalidate all the sharding prop caches. Disable it for now.

cc H-Huang awgu wanchaol fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

wconstab reviewed (approved):
lgtm. let me know if the cache clearing thing ends up being a blocker - (basically, is it an important use case to let someone deregister CP after using it?)

wconstab commented on torch/distributed/tensor/experimental/_context_parallel/_sharding_rules.py:
```diff
@@ -0,0 +1,399 @@
+            propagator.op_to_schema_info[op_overload] = _origin_op_strategy_schema
+
+        # Clear cache
+        propagator.propagate_op_sharding.cache.cache_clear()
```
oh- note- this way of clearing the cache is no longer correct.  The cache that you would need to clear is the one in c++ (python_variable.cpp).  This `propagate_op_sharding` LRU cache is basically not used anymore by dtensor.

(it appears to still be used by tensor-parallel /convolution handler, which is the only reason it is not deleted)

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update on "[CP] Refactor CP sharding rules into separate module and only register when CP is enabled"


Previously, CP-specific sharding strategies (which shard on the sequence dimension) were directly
included in the base sharding strategies for scaled_dot_product_attention operators in
`_matrix_ops.py`. This meant these strategies were always available, even when CP was not enabled,
which could lead to incorrect sharding behavior as these sharding rules are not mathmetically correct without CP.

1. **Created new module**:
`torch/distributed/tensor/experimental/_context_parallel/_sharding_rules.py`
   - Implements `op_strategy_context()` - a context manager for temporarily
registering/unregistering strategies
   - Defines CP-enhanced strategy functions for all 6 scaled_dot_product_attention ops (forward and
 backward for flash, efficient, and cudnn variants)
   - Provides `register_cp_sharding_rules()` and `unregister_cp_sharding_rules()` APIs

2. **Updated `_matrix_ops.py`**
   - Removed all CP-specific sharding rules (sequence dimension sharding strategies)
   - Base strategies now only contain replicate, tensor parallelism, and batch sharding
strategies

3. **Updated `_attention.py`**
   - `_enable_cp_dtensor_dispatcher()` now calls `register_cp_sharding_rules()` to dynamically add
CP strategies
   - ~`_disable_cp_dtensor_dispatcher()` now calls `unregister_cp_sharding_rules()` to restore
original strategies~ This will invalidate all the sharding prop caches. Disable it for now.

cc H-Huang awgu wanchaol fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]

fegin committed:
Update

[ghstack-poisoned]
